### DISCUSSION TITLE
Rename `empty` operator

### DISCUSSION THREAD

**Discussion by sergeysova** [sergeysova/patronum#224] [General]
Looks like "empty" semantically should check for empty values of any kind.
We have two options: 1) Change operator behavior 2) rename the operator

Examples:
- https://ramdajs.com/docs/#isEmpty
- https://lodash.com/docs/4.17.15#isEmpty
- https://doc.rust-lang.org/stable/std/primitive.str.html#method.is_empty
- https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.is_empty

Available candidates for the new name are written down. Post your candidates and vote for another

**Comment by sergeysova** [comment]
`absent`

**Comment by sergeysova** [comment]
`nil`

**Reply by demark** [reply]
but this one should check against both `null` or `undefined` (see https://lodash.com/docs/#isNil).

**Reply by igorkamyshev** [reply]
> but this one should check against both `null` or `undefined` (see https://lodash.com/docs/#isNil).

It's impossible to have 'undefined' in store.

**Reply by demark** [reply]
Good point. Implementation-wise `isNil` is usually as simple as `value == null` check which covers both types and would fit for Effector cases as well. Eg. `sample({ clock, filter: not(nil($something)), ... })` looks reasonable and concise to me.

**Comment by sergeysova** [comment]
`isEmpty`

**Comment by sergeysova** [comment]
`empty`.

As is. Do not change

**Comment by yumauri** [comment]
`isNull`

**Comment by igorkamyshev** [comment]
In my opinion, `empty` semantics in the mentioned packages is broken. `null` is special value on our platform to mark some value as intentionally absent. Empty strings, arrays and maps without elements are specific to a particular application and cannot be named "empty" in any case.

So, it is strictly correct to name `null`-value as "empty". I suggest to keep current naming for `empty`-operator and introduce new operator `size` to check mentioned cases.

```ts
const $noElements = equals(size($arrStore), 0);
const $emptyString = equals(size($stringStore), 0);
const $noPairs = equals(size($mapStore), 0);
```

**Reply by yumauri** [reply]
Wild suggestions to add some more FP-ish style? 🤪
```js
const zeroed = equals(0) // partially applied

const $noElements = zeroed(size($arrStore))
const $emptyString = zeroed(size($stringStore))
const $noPairs = zeroed(size($mapStore))

const valued = not(zeroed) // partially applied

const $hasElements = valued(size($arrStore))
const $hasString = valued(size($stringStore))
const $hasPairs = valued(size($mapStore))
```
But this could lead to a huge mess with types, I guess ._.

**Comment by nvmxre** [comment]
`isVoid`

**Comment by kirillku** [comment]
I would keep it named `empty` to be consistent with the rest of operators (we don't have any `isSomething`, always just `something`). But i would check also for `""`, `[]`, `{}`, so it would be analog of doing `$store.map(_.isEmpty)`.

**Reply by pijng** [reply]
In ruby, for example, there is a `blank` method to check for empty arrays, strings, objects and null as well.

**Comment by Kelin2025** [comment]
**Change behavior to match `isEmpty` from other libraries to avoid confusion**  

We already have [people](https://github.com/effector/effector/discussions/737) [complaining](https://github.com/effector/effector/issues/721) about specific naming/terms used in Effector and now we want to re-define `empty`. Not good

**Comment by sergeysova** [comment]
`blank`

**Reply by igorkamyshev** [reply]
> `blank`

Shall we consider the following options?
- lol
- kek
- pek

😂

**Reply by sergeysova** [reply]
@igorkamyshev 

https://apidock.com/rails/Object/blank%3F

https://blog.arkency.com/2017/07/nil-empty-blank-ruby-rails-difference/

**Reply by igorkamyshev** [reply]
Luckily we are not using RoR ☺️

**Comment by sergeysova** [comment]
`nullish`

**Comment by ifeelBALANCED** [comment]
@sergeysova 

`isVacant`
`isUnoccupied`
`isBlank`
`isDevoid`
`nullify`
`wipeOut`
`annihilate`
